### PR TITLE
Added go to implementation service

### LINKED ISSUE #598: Support `textDocument.implementation` LSP request
See [the spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_implementation) for more info on the request type.

### PR DESCRIPTION
**Pull request by gfontorbe**
Added abstract service for handling go to implementation requests

Closes #598

### CODE REVIEW

File: packages/langium/src/services.ts [hunk 1]
```diff
@@ -47,6 +47,7 @@ import type { WorkspaceManager } from './workspace/workspace-manager';
 import type { LanguageServer } from './lsp/language-server';
 import type { SignatureHelpProvider } from './lsp/signature-help-provider';
 import type { GoToTypeProvider } from './lsp/goto-type';
+import { GoToImplementationProvider } from './lsp/goto-implementation';
```
Comment (msujew) [inline packages/langium/src/services.ts:50]: Minor: Please use `import type` here.

File: packages/langium/src/lsp/goto-implementation.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    protected readonly references: References;
+
+    constructor(services: LangiumServices) {
+        this.references = services.references.References;
+    }
+
+    goToImplementation(document: LangiumDocument<AstNode>, params: ImplementationParams, cancelToken = CancellationToken.None): MaybePromise<LocationLink[] | undefined> {
+        const rootNode = document.parseResult.value;
+        if (rootNode.$cstNode) {
+            const sourceCstNode = findLeafNodeAtOffset(rootNode.$cstNode, document.textDocument.offsetAt(params.position));
```
Comment (msujew) [inline packages/langium/src/lsp/goto-implementation.ts:35]: Suggestion: Please use `findDeclarationNodeAtOffset` instead.

**Review by msujew — COMMENTED** [review]
One minor thing, afterwards this is good to go 👍

**Comment by gfontorbe** [comment]
Looks like I did something wrong with the rebase, any way to clean that up?

**Review by msujew — APPROVED** [review]
Looks good to me 👍

**Pull request merged**